## Logistic Regression with boston data

In [1]:
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

### Data manipulation

In [2]:
from numpy import genfromtxt
X = genfromtxt('./data/X_ber.csv', delimiter=',',skip_header=1)
y = genfromtxt('./data/y_ber.csv', delimiter=',',skip_header=1)

In [3]:
X[0:5, :] # numpy array
# x1 : gpa, x3: rank of school

array([[  1.  , 380.  ,   3.61,   3.  ],
       [  1.  , 660.  ,   3.67,   3.  ],
       [  1.  , 800.  ,   4.  ,   1.  ],
       [  1.  , 640.  ,   3.19,   4.  ],
       [  1.  , 520.  ,   2.93,   4.  ]])

In [4]:
print(X.shape, y.shape)

(400, 4) (400,)


In [ ]:
# careful of broadcasting

## My first Logistic Regression Model

#### These are some example of latex formula

$ \frac{1}{3} \sqrt{555} \ sin $

$y^{(i)} \sim {\rm Ber}(p^{(i)})$ where
$ p^{(i)}=\sigma(X^{(i)}\beta)$
and $\sigma(t)=\frac{1}{1+exp(t)}$

$ $

In [4]:
# Calculate beta?
# 1. Define the model
glm_logistic = sm.GLM(y, X, family=sm.families.Binomial())
# stats model -> sm + .NN, .Tree
# response v + explanatory v + family : logistic regression (y~ Ber), p : default sigmoid

logistic_results = glm_logistic.fit(); # beta

In [5]:
print(logistic_results.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                      y   No. Observations:                  400
Model:                            GLM   Df Residuals:                      396
Model Family:                Binomial   Df Model:                            3
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -229.72
Date:                Wed, 11 Nov 2020   Deviance:                       459.44
Time:                        17:34:44   Pearson chi2:                     399.
No. Iterations:                     4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.4495      1.133     -3.045      0.0

In [ ]:
# only coef -> beta

$beta.hat$ = logistic parameter ?

$\hat {\beta} = (-3.4495, 0.0023, 0.7770, -0.5600)$

Next goal : Calculation of $p^{(i)}$

In [14]:
beta_hat=logistic_results.params # 위 값과 동일
temp=np.dot(X, beta_hat) # = X*beta
p_hat = 1/(1+np.exp(-temp))
p_hat[0:5]
# accepted probability
# You can do the same thing the following for the calculation of p_hat

array([0.18955274, 0.31778074, 0.71781361, 0.14894919, 0.0979542 ])

In [11]:
yhat = logistic_results.mu # prediction for each observation
# mean indicates probability

In [19]:
yhat[0:5] # 동일한 결과

array([0.18955274, 0.31778074, 0.71781361, 0.14894919, 0.0979542 ])

* Calculate the prediction error!

In [15]:
admission_hat = yhat>0.5 # dummy v
np.mean(admission_hat==y)

0.705

In [ ]:
admission_hat[0:5]

In [17]:
y[0:5]

array([0., 1., 1., 1., 0.])

* Calculate the prediction error again!

In [16]:
W = logistic_results.params # Estimated parameters
my_y_hat = 1/(1+np.exp(-np.dot(X, W)))
my_admission_hat = my_y_hat>0.5
np.mean(my_admission_hat==y)

0.705